### PR TITLE
Add support for parentTemplate calls, similar to parentData.

### PR DESCRIPTION
**Pull request by queso**
I talked to @stubailo about this a little. I used this so I could create a search/filter type component using a reactive var instead of needing messy/global session vars. I can imagine there are many other use cases for this type of function call being available.

Tests are included, as well as some lightweight docs.

Do you guys have a CI server setup that shows the status? I have to comment out a good bit of tests to even get things running.

### CODE REVIEW

**Comment by glasser** [comment]
@stubailo can you respond to this one?

BTW, what do you mean about tests?  Does `./meteor test-packages` not pass for you?

**Comment by queso** [comment]
No, I found the tests didn't run at all and found four failing tests.

**Comment by typ90** [comment]
+1

**Comment by queso** [comment]
@glasser any update on this? Would love to make sure that all the blaze tests are passing. Is there a public CI showing the build anywhere? Or just a private jenkins server somewhere?

**Comment by amandasaffer** [comment]
+1

**Comment by shilman** [comment]
+1

File: packages/blaze/template.js [hunk 1]
```diff
@@ -431,6 +431,26 @@ Blaze.TemplateInstance.prototype.subscriptionsReady = function () {
   return this._allSubsReady;
 };
 
+
+/**
+ * @summary Accesses other template instances that enclose the current template instance.
+ * @locus Client
+ * @function
+ * @param {Integer} [numLevels] The number of levels beyond the template instance to look. Defaults to 1.
```
Comment (shilman) [inline packages/blaze/template.js]: `numLevels` => `levels`?

**Comment by rclai** [comment]
Can you make it so that you can also pass in a `function`:

```js
var parentTemplateInstance = Template.instance().parentTemplate(function (parentInstance) {
  return parentInstance.view.name === 'someTemplateName';
});
```

File: packages/blaze/template.js (hunk 1 above)
Comment (queso) [inline packages/blaze/template.js]: Thx, will fix.

**Comment by queso** [comment]
@rclai sure, I can try to work that in.

It still puzzles me, having created this PR with tests, how we get this integrated into Blaze or what the next steps are?

**Comment by glasser** [comment]
I'm not sure what you mean about tests not passing. I see all test-packages tests passing when I check out your PR. Which tests don't pass for you?

I'm not reviewing the PR itself (since you said this is something you discussed with @stubailo, I assume he has more context) but I am curious about why you think tests aren't passing.

**Comment by queso** [comment]
I had to comment out **other** tests to get test-packages to run, looked like stuff around let and other upcoming blaze changes.

I am aware that my tests pass ;)

I will see if I can follow up with @stubailo about this.

**Comment by glasser** [comment]
A clean checkout of `pull/4449` passes all `test-packages` tests.

You have now asserted four times on this issue that other tests are failing for you but haven't even mentioned the name of one of those failing tests, shown the failure, etc.  https://github.com/meteor/meteor/wiki/Contributing-to-Meteor has some advice about how to share bug reports, and "tests are failing on devel now" is a bug report.

**Comment by queso** [comment]
I had to remove these lines from templates to get tests to run for blaze:

```
diff --git a/packages/spacebars-tests/template_tests.html b/packages/spacebars-tests/template_tests.html
index 36e8fa9..5d8c888 100644
--- a/packages/spacebars-tests/template_tests.html
+++ b/packages/spacebars-tests/template_tests.html
@@ -1101,16 +1101,13 @@ Hi there!

 <template name="spacebars_template_test_each_index">
   {{#each things}}
-    <span>{{@index}} - {{num}}</span>
   {{/each}}
 </template>

 <template name="spacebars_template_test_nested_exprs">
-  {{add (add 1 2) 3}}
 </template>

 <template name="spacebars_template_test_nested_subexprs">
-  {{capitalize (firstWord generateSentence)}}
 </template>

 <template name="spacebars_template_test_exprs_keyword_callable">
@@ -1118,7 +1115,6 @@ Hi there!
 </template>

 <template name="spacebars_template_test_exprs_keyword">
-  {{> spacebars_template_test_exprs_keyword_callable stuff=(capitalize name) another=(capitalize "mello")}}
 </template>
```

**Comment by queso** [comment]
@glasser, maybe I am doing something wrong here, but I just checked out `devel` and ran `meteor test-packages ./packages/spacebars-tests` and I can't get it to run because of this error message:

```
=> Started proxy.
=> Started MongoDB.
=> Errors prevented startup:

   While building package local-test:spacebars-tests:
   template_tests.html:1094: Expected IDENTIFIER
   ...things}}     <span>{{@index}} - {{num}}</...
   ^

=> Your application has errors. Waiting for file change.
```

How should I be running the test-packages, I don't see anything about that in the contributing file you linked to. Further, if the CI server were somehow public or you documented the testing process somewhere, I would be happy to ensure I not bother anyone with these questions.

I would also love to see tests run for PRs? That would at least give me some insight that my test suite updates are good to go for you guys :)

**Comment by glasser** [comment]
Are you running `./meteor` or your installed meteor? Yes, taking your installed meteor and running it explicitly against the devel tests might not work.

**Comment by queso** [comment]
Ugh. ┬─┬﻿ ︵ /(.□. \）

You were right, I was running `meteor` and not `./meteor`. I will create a seperate issue for the Contrib doc to be updated perhaps to save others my pain and suffering, haha.

**Comment by dandv** [comment]
Any updated to the [Contributing guidelines](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor) about running `./meteor test-packages` properly?

**Comment by queso** [comment]
@dandv I'm not really sure I can update those as they are part of the wiki. I could write some markdown and put it here for adding, but I don't think I can submit a PR for a wiki commit.

**Comment by queso** [comment]
Ok, I am glad to see that CI has been added to the build process now for PRs. I have rebased my commit to be up to date with the latest devel branch so we can get those tests up to date for CI. I also address @shilman's concern of mismatching variables and documentation names.

@glasser and @danv: I have written up a bit on running tests that I think should be added to the Contribute guide wiki: https://meteor.hackpad.com/Running-Meteor.js-test-suite-4dnr0vgcfrv

**Comment by glasser** [comment]
Just to be clear: I don't really have any opinions about the feature in this PR; I am not really a Blaze expert, and mostly only paid attention to this PR because of the claim that test-packages was failing on devel.

I agree that your suggestions would be a great improvement to the Contributing wiki!  I've personally been so focused on Galaxy lately and not really in the loop around Meteor framework contributions; @stubailo, want to add something like that to the Contributing page?  It doesn't look like there are docs *anywhere* telling you how to run the tests!

File: packages/blaze/template.js [hunk 2]
```diff
@@ -431,6 +431,26 @@ Blaze.TemplateInstance.prototype.subscriptionsReady = function () {
...
+ */
+Blaze.TemplateInstance.prototype.parentTemplate = function (numLevels) {
+    var view = Blaze.currentView;
+    if (typeof numLevels === "undefined") {
+        numLevels = 1;
+    }
+    while (view) {
+        if (view.name.substring(0, 9) === "Template." && !(numLevels--)) {
```
Comment (mitar) [inline packages/blaze/template.js]: Uhu. This is really bad. Using view name to determine the nature of the view? This is not how it should be done.

**Comment by mitar** [comment]
OK, as an author of the similar code in [template-extension](https://github.com/aldeed/meteor-template-extension) package, I think the approach taken here is at the current state not the best one and should not be merged in as it is.

Please look at [the implementation here](https://github.com/aldeed/meteor-template-extension/blob/master/template-extension.js#L259). Feel free to simply use it and submit it in the pull request. It is production tested and working on many sites. And does not bring any assumptions about naming of views.

I would also suggest that the function is simply called `parent`. Because in the code is then common to have `template.parent()`. It reads pretty naturally. Also, it would make all the code which is currently using `template-extension` magically work.

So in short, what I am probably saying is that if MDG is open to such feature, probably the best way would be to merge the implementation from `template-extension` package in. (See there also for tests.)

**Comment by stubailo** [comment]
OK, sounds good. As I understand there is already an issue for merging in template-extension, right?

Could it simply be a matter of documenting `template-extension` in a visible place?

**Comment by mitar** [comment]
That might be a wider question of MDG endorsing some 3rd party libraries and listing them? Like the set of libraries one might want to look into if they start working on the project? (It would be also great if MDG would spend a bit of time reviewing each of those and giving some feedback.)

But for this particular feature I think it is small enough to be directly included in the core.

**Comment by queso** [comment]
Since you closed my PR, could you perhaps at least point to the other issue @stubailo? Again, kind of frustrating to just see a PR closed with pretty much 0 discussion from MDG.

**Comment by mitar** [comment]
See #3706.

**Comment by stubailo** [comment]
Hey sorry - for some reason I thought it was Mitar's PR. And yes, recommending non-core packages for people to use is coming ultra soon. I hope to get started on it next week, with ample community involvement from all of you. Hang tight, just ironing out some ideas.

File: packages/blaze/template.js (hunk 2 above)
Comment (gabrielhpugliese) [inline packages/blaze/template.js]: :+1:

**Comment by queso** [comment]
@mitar looks like #3706 isn't really progressing to me. I would prefer to keep this open and work the goal of this PR into core.

@stubailo no worries, glad it was a mistake and not an outright denial of adding the feature :)

**Comment by mitar** [comment]
@queso: I completely agree that we might want to put this in step by step. I am just encouraging you to take the implementation from template-extension and use it here. :-) And then let's merge this. :-)

**Comment by tmeasday** [comment]
Thanks for this PR, sorry we never got around to merging it. We are currently cleaning up the repository and removing old PRs; if you think this PR still has value and are interested in resurrecting it, we encourage you to re-submit it against https://github.com/meteor/blaze

**Comment by queso** [comment]
Appreciate the official update, but I doubt there is any value here for me
to submit this again. I have personally moved away from Blaze and to the
more performant option of React. The client that I did this for, I just
ended up monkey patching in what I needed anyway.
